src/client/driver/command-executors/client-functions/selector-executor/index.js

```diff
@@ -63,6 +64,9 @@ export default class SelectorExecutor extends ClientFunctionExecutor {
                 const isElementVisible   = !this.command.visibilityCheck || visible(el);
                 const isTimeout          = new DateCtor() - startTime >= this.timeout;
 
+                if (domUtils.isShadowRoot(el))
+                    throw new ActionElementIsShadowRootElementError();
```

Comment:
I don't feel that here is the right place for this check. By doing it here, we prevent using expressions like `t.expect(element.shadowRoot()).contains('text')` that are perfectly valid in my opinion. I think it's better to do it here:
https://github.com/DevExpress/testcafe/blob/master/src/client/driver/command-executors/execute-action.js#L157